AGENTS.md

```diff
@@ -0,0 +1,293 @@
+# Insights MCP - AI Coding Assistant Guide
+
+This document provides AI coding assistants with specific guidance for working with the Insights MCP codebase. **Start by reading [README.md](README.md)** for project overview, authentication setup, and client integrations.
+
+This guide supplements the README with development-specific information and workflow guidance for AI coding assistants.
+
+## Architecture
+
+### Core Components
+
+1. **InsightsMCPServer** (`src/insights_mcp/server.py`)
+   - Main unified server that mounts multiple service toolsets
+   - Handles authentication and configuration for all mounted toolsets
+   - Supports selective toolset registration via `--toolset` argument
+
+2. **InsightsMCP Base Class** (`src/insights_mcp/mcp.py`)
+   - Abstract base class for all Insights MCP toolsets
+   - Manages authentication and client initialization
+   - Provides common functionality for Red Hat Insights API integration
+
+3. **InsightsClient** (`src/insights_mcp/client.py`)
+   - HTTP client for Red Hat Insights APIs with OAuth2 support
+   - Handles service account and refresh token authentication
+   - Provides error handling and proxy support
+
+4. **OAuth Middleware** (`src/insights_mcp/oauth.py`)
+   - Starlette middleware for OAuth flows in HTTP/SSE transports
+   - Dynamic client registration and metadata proxying
+
+### Toolset Architecture
+
+The project uses a **toolset-based architecture** where each service is implemented as a separate toolset:
+
+- **image-builder** (`src/image_builder_mcp/server.py`): Image Builder service tools
+- **example** (`src/insights_mcp/servers/example.py`): Example toolset for development
+
+**Toolset Registration:**
+- Toolsets are registered in `MCPS` list in `server.py`
+- Each toolset has a unique `toolset_name` for identification
+- Tools are mounted with toolset prefix (e.g., `image-builder_get_blueprints`)
+- Users can select specific toolsets: `insights-mcp --toolset=image-builder`
+
+### Transport Modes
+
+- **STDIO** (default): Direct process communication for desktop integrations
+- **HTTP**: RESTful API with streaming capabilities  
+- **SSE**: Server-sent events for real-time web clients
+
+## Development Workflow for AI Assistants
+
+### Quick Setup
+```bash
+# Prerequisites: Python 3.10+, uv package manager
+uv venv && source .venv/bin/activate
+make install-test-deps  # Installs all dev dependencies
+```
+
+### Essential Commands
+```bash
+make help  # Show all available make targets
+```
+
+**Key development commands:**
+- `make test` - Run test suite
+- `make lint` - Run all linting
+- `make build` - Build container image
+- `insights-mcp` - Run server in development mode
+
+**Note**: Commands require `source .venv/bin/activate` first.
+**Note**: Authentication setup is covered in [README.md](README.md).
+
+## Testing
+
+### Test Structure
```

Comment:
I think this could be improved to provide a few other examples now, but also just to mention the pattern of where to find tests.